docs/dataset/advanced-usage/set-up-natural-language-search.md

```diff
@@ -0,0 +1,148 @@
+This is used by the uploader for authentication against your Kolena instance.
+
+```shell
+export KOLENA_TOKEN="********"
+```
+
+Configure an additional poetry source:
+```shell
+poetry source add --priority=supplemental kolena-embeddings "https://gateway.kolena.cloud/repositories"
+```
+
+Then run the following command, making sure to replace <KOLENA_TOKEN> with the token retrieved from the developer page:
+```shell
+poetry config http-basic.kolena-embeddings <KOLENA_TOKEN> ""
+```
+
+This package provides the `kembed.util.extract_embeddings` method that generates
+embeddings as a numpy array for a given [`PIL.Image.Image`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image)
+object.
+
+### Step 2: Load Dataset and Model
+
+Before extracting embeddings on a dataset, we need to load the dataset. The dataset
+seeded in the [semantic_segmentation](https://github.com/kolenaIO/kolena/tree/trunk/examples/dataset/semantic_segmentation)
+example contains image assets referenced by the `locator`
+column, and we load the dataset in to a dataframe.
+
+The embedding model and its key are obtained via the `load_embedding_model()` method.
+
+```python
+from kembed.util import load_embedding_model
+from kolena.dataset import download_dataset
+
+df_dataset = download_dataset("coco-stuff-10k")
+model, model_key = load_embedding_model()
+```
+
+### Step 3: Load Images for Extraction
+
+In order to extract embeddings on image data, we must load our image files into a
+[`PIL.Image.Image`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image) object.
+In this section, we will load these images from an S3 bucket.
+For other cloud storage services, please refer to your cloud storage's API docs.
+
+```python
+from typing import Iterator
+from typing import List
+from typing import Tuple
+
+import boto3
+from PIL import Image
+
+s3 = boto3.client("s3")
+
+def load_image_from_accessor(accessor: str) -> Image:
+    bucket_name, *parts = accessor[5:].split("/")
+    file_stream = boto3.resource("s3").Bucket(bucket_name).Object("/".join(parts)).get()["Body"]
+    return Image.open(file_stream)
+
+def iter_image_paths(image_accessors: List[str]) -> Iterator[Tuple[str, Image.Image]]:
+    for locator in image_accessors:
+        image = load_image_from_accessor(locator)
+        yield locator, image
+```
+
+!!! tip end
+    When processing large scales of images, we recommend using an `Iterator` to limit the number
+    of images loaded into memory at once.
+
+### Step 4: Extract and Upload Embeddings
+
+Once embeddings are extracted for each `locator` on the dataset, we create a dataframe with
+`embedding` and `locator` columns, and use the `upload_dataset_embeddings` method to upload
+the embeddings.
+
+The dataframe uploaded is required to contain the ID columns of the dataset in order to
+match against the [`Datapoints`](../core-concepts/dataset.md/#datapoints) in the dataset.
```

Comment:
Nit: since it's linking to a section rather than an API definition, lowercase and normal font for "datapoints"